pkg/migration-tool/planner/block.go

```diff
@@ -25,32 +25,103 @@ type Block struct {
+}
+
+// Fetch starts fetching the samples from remote read storage based on the matchers. It takes care of concurrent pulls as well.
+func (b *Block) Fetch(ctx context.Context, client *utils.Client, mint, maxt int64, matchers []*labels.Matcher) (containsTs bool, err error) {
+	var (
+		totalRequests         = b.numStores
+		readRequests          = make([]*prompb.Query, totalRequests)
+		readResponses         = make([]*prompb.QueryResult, totalRequests)
+		contexts              = make([]context.Context, totalRequests)
+		cancelFuncs           = make([]context.CancelFunc, totalRequests)
+		timeRangeMinutesDelta = (maxt - mint) / time.Minute.Milliseconds()
+	)
+	for i := 0; i < totalRequests; i++ {
+		readRequest, err := utils.CreatePrombQuery(b.stores[i].mint, b.stores[i].maxt, matchers)
+		if err != nil {
+			return false, fmt.Errorf("create promb query: %w", err)
+		}
+		readRequests[i] = readRequest
+		cctx, cancelFunc := context.WithCancel(ctx)
+		contexts[i] = cctx
+		cancelFuncs[i] = cancelFunc
+	}
+	var (
+		doneChan = make(chan *utils.PrompbResponse, totalRequests)
+		errChan  = make(chan error, totalRequests)
+	)
+	for i := 0; i < totalRequests; i++ {
+		// Initiate concurrent fetching of data.
+		descriptionMsg := fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta)
+		go client.Read(contexts[i], readRequests[i], i, descriptionMsg, doneChan, errChan)
+	}
+	var (
+		bytesCompressed   int
+		bytesUncompressed int
+	)
+	for i := 0; i < totalRequests; i++ {
+		select {
+		case resp := <-doneChan:
+			bytesCompressed += resp.NumBytesCompressed
+			bytesUncompressed += resp.NumBytesUncompressed
+			readResponses[resp.ID] = resp.Result
+		case err = <-errChan:
+			for _, cancelFnc := range cancelFuncs {
+				cancelFnc()
+			}
+			return false, fmt.Errorf("executing client-read: %w", err)
+		}
 	}
-	result, bytesCompressed, bytesUncompressed, err := client.Read(context, readRequest, fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta))
-	if err != nil {
-		return nil, fmt.Errorf("executing client-read: %w", err)
+	close(doneChan)
+	close(errChan)
+	// Combine the stores.
+	var result []*prompb.TimeSeries
+	for storeIndex := 0; storeIndex < b.numStores; storeIndex++ {
+		result = append(result, readResponses[storeIndex].Timeseries...)
 	}
-	b.timeseries = result.Timeseries
+	b.timeseries = result
 	// We set compressed bytes in block since those are the bytes that will be pushed over the network to the write storage after snappy compression.
 	// The pushed bytes are not exactly the bytesCompressed since while pushing, we add the progress metric. But,
 	// the size of progress metric along with the sample is negligible. So, it is safe to consider bytesCompressed
 	// in such a scenario.
 	b.numBytesCompressed = bytesCompressed
 	b.numBytesUncompressed = bytesUncompressed
 	b.plan.update(bytesUncompressed)
-	return result, nil
+	return len(result) != 0, nil
```

Comment:
how is this different from len(b.timeseries) != 0 and if it's not why do we need a separate return?

Reply:
The only reason why earlier we used to return result complete is it was being used to check the length of the result. If the length is 0, decrement the block count. That's it. This simple logical operation can be done here itself and hence make this more efficient by decreasing the functional return value.

Reply:
sorry but we already have b.timeseries which is what we need for the check. We don't actually need neither result not len(result)!=0.

Reply:
Yeah I agree, that would be very efficient. The b.timeseries thing. I have updated accordingly. Thank you.